libkpod/attach.go

```diff
@@ -0,0 +1,182 @@
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+)
+
+/* Sync with stdpipe_t in conmon.c */
+const (
+	AttachPipeStdin  = 1
+	AttachPipeStdout = 2
+	AttachPipeStderr = 3
+)
+
+// ContainerAttach attaches to a running container
+// keys are string representation of what to use to detach
+// key format is a single character [a-Z] or ctrl-<value> where <value> is one of: a-z, @, ^, [, , or _.
+func (c *ContainerServer) ContainerAttach(container string, noStdIn bool, keys string) error {
+	// Check the validity of the provided keys first
+	var err error
+	detachKeys := []byte{}
+	if len(keys) > 0 {
+		detachKeys, err = term.ToBytes(keys)
+		if err != nil {
+			return errors.Wrapf(err, "invalid detach keys")
+		}
+	}
+
+	ctr, err := c.LookupContainer(container)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find container %s", container)
+	}
+
+	cStatus := c.runtime.ContainerStatus(ctr)
+	if !(cStatus.Status == oci.ContainerStateRunning || cStatus.Status == oci.ContainerStateCreated) {
+		return errors.Errorf("%s is not created or running", container)
+	}
+
+	resize := make(chan remotecommand.TerminalSize)
+
+	AttachContainerSocket(ctr, resize, noStdIn, detachKeys)
+	close(resize)
+	c.ContainerStateToDisk(ctr)
+
+	return nil
+}
+
+// getStandardStreams returns the standard IO readers/writers
+func getStandardStreams() (stdIn, stdOut, stdErr *os.File) {
+	return os.Stdin, os.Stdout, os.Stderr
+}
+
+func closeInputStream(stdIn io.ReadCloser) { //nolint
+	stdIn.Close()
+}
+
+// AttachContainerSocket connects to the container's attach socket and deals with the IO
+func AttachContainerSocket(ctr *oci.Container, resize <-chan remotecommand.TerminalSize, noStdIn bool, detachKeys []byte) error {
+	inputStream, outputStream, errorStream := getStandardStreams()
+	defer closeInputStream(inputStream)
+
+	tty, err := strconv.ParseBool(ctr.State().Annotations["io.kubernetes.cri-o.TTY"])
+	if err != nil {
+		return errors.Errorf("unable to parse annotations in %s", ctr.ID)
+	}
+	if !tty {
+		return errors.Errorf("no tty available for %s", ctr.ID())
+	}
+
+	oldTermState, err := term.SaveState(inputStream.Fd())
+	defer term.RestoreTerminal(inputStream.Fd(), oldTermState)
+
+	if err != nil {
+		return errors.Wrapf(err, "unable to save terminal state")
+	}
+
+	// Put both input and output into raw
+	if !noStdIn {
+		term.SetRawTerminal(inputStream.Fd())
+	}
+
+	controlPath := filepath.Join(ctr.BundlePath(), "ctl")
+	controlFile, err := os.OpenFile(controlPath, unix.O_WRONLY, 0)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open container ctl file: %v")
+	}
+
+	kubecontainer.HandleResizing(resize, func(size remotecommand.TerminalSize) {
+		logrus.Debug("Got a resize event: %+v", size)
+		_, err := fmt.Fprintf(controlFile, "%d %d %d\n", 1, size.Height, size.Width)
+		if err != nil {
+			logrus.Warn("Failed to write to control file to resize terminal: %v", err)
+		}
+	})
+	attachSocketPath := filepath.Join("/var/run/crio", ctr.ID(), "attach")
```

Comment:
/var/run/crio should not be hard coded, should be gotten from oci.Runtime or some where where it can be customized for running inside of a test framework.

Reply:
Im going to file an issue for this one ... it's a little broader than this PR.

Reply:
Can't you just get it out of the config?

// DefaultConfig returns the default configuration for crio.
func DefaultConfig() *Config {
        return &Config{
                RootConfig: RootConfig{
                        Root:        crioRoot,
                        RunRoot:     crioRunRoot,
                        LogDir:      "/var/log/crio/pods",
                        FileLocking: true,
                },
config.RootConfig.RunRoot?

Reply:
Need to make the same change to CRI-O to make this work, at least for now.

Reply:
I'm not opposed to doing it, but if we do, we should do it here and in the original code.

Reply:
Yes I guess if the original code has that hard coded we need to change it there also.

Reply:
Hence why I recommended we handle this in a separate issue.  Ok with that?

Reply:
Sure although not sure why.  I will open a PR for fixing it in crio.

Reply:
Ok I was wrong. RunRoot is for containers/stoarage not for the /var/run directory.  Seems we have no way of changing that so leave this alone.

Reply:
baude create it as a constant, and then we can use it in the other code.  It should be defined in libpod not used directly in the crio code.